### ISSUE TITLE
conv2d_transpose produce different results on GPU

### ISSUE THREAD

**Issue by jnjaby** [tensorflow/tensorflow#11882]
### System information
== cat /etc/issue ===============================================
Linux ST 4.2.0-42-generic #49~14.04.1-Ubuntu SMP Wed Jun 29 20:22:11 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
VERSION="14.04.4 LTS, Trusty Tahr"
VERSION_ID="14.04"

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 4.9.4-2ubuntu1~14.04.1) 4.9.4
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

== uname -a =====================================================
Linux ST 4.2.0-42-generic #49~14.04.1-Ubuntu SMP Wed Jun 29 20:22:11 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
msgpack-numpy (0.4.1)
numpy (1.13.1)
protobuf (3.2.0)
tensorflow (0.10.0)
tensorflow-gpu (1.0.0)

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.0.0
tf.GIT_VERSION = v1.0.0-rc2-15-g47bba63-dirty
tf.COMPILER_VERSION = v1.0.0-rc2-15-g47bba63-dirty
Sanity check: array([1], dtype=int32)
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.so.5 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.so.8.0 locally

== env ==========================================================
LD_LIBRARY_PATH /home/abc/torch/install/lib:/usr/lib/x86_64-linux-gnu:/home/abc/torch/install/lib:/usr/local/cuda/lib64:/usr/local/cuda/lib64:/home/abc/torch/install/lib:/home/abc/code/torch/torch/install/lib:/usr/local/cuda/lib64::/usr/local/computecpp/lib:/data/software/gurobi652/linux64/lib
DYLD_LIBRARY_PATH /home/abc/torch/install/lib:/home/abc/torch/install/lib:/home/abc/code/torch/torch/install/lib:

== nvidia-smi ===================================================
Sun Jul 30 17:45:45 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.48                 Driver Version: 367.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1080    Off  | 0000:01:00.0      On |                  N/A |
|  0%   53C    P2    47W / 260W |   7909MiB /  8112MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      1308    G   /usr/bin/X                                     357MiB |
|    0      2590    G   compiz                                         229MiB |
|    0      3254    G   ...el-token=CBAE43C38254E155E78826C3F38F0092    99MiB |
|    0      9480    C   python                                        1039MiB |
|    0     10432    C   /usr/bin/python                               5895MiB |
|    0     20408    C   /usr/bin/python                                283MiB |
|    0     28024    G   /usr/local/MATLAB/R2015a/bin/glnxa64/MATLAB      2MiB |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/lib/python2.7/dist-packages/torch/lib/libcudart.so.8.0
/usr/local/lib/python2.7/dist-packages/torch/lib/libcudart.so
/usr/local/cuda-8.0/lib64/libcudart.so.8.0.44
/usr/local/cuda-8.0/lib64/libcudart_static.a
/usr/local/cuda-8.0/doc/man/man7/libcudart.7
/usr/local/cuda-8.0/doc/man/man7/libcudart.so.7
/usr/local/MATLAB/R2017a/bin/glnxa64/libcudart.so.8.0.44
/usr/local/MATLAB/R2015a/bin/glnxa64/libcudart.so.6.5.14
### Describe the problem
I am trying to use `tf.nn.conv2d_transpose` but it produces different results every time on GPU. However, the result would be the same when switching the device to CPU. It seems like a bug. Please check the toy model below for more details.

### Source code / logs
```python
import tensorflow as tf
import numpy as np

np.random.seed(1234)
conv_ = np.random.randn(10, 7, 7, 56)

with tf.device('/gpu:0'):
    bottom = tf.constant(conv_, dtype=tf.float32)
    weight = tf.get_variable("weight", [9, 9, 1, 56], initializer=tf.random_normal_initializer(0, 0.001))
    bias = tf.get_variable("bias", initializer=np.zeros(1, dtype=np.float32))	

    conv = tf.nn.conv2d_transpose(bottom, weight, [10, 19, 19, 1], [1, 3, 3, 1], padding='SAME')
    conv = tf.nn.bias_add(conv, bias)

sess = tf.Session()
sess.run(tf.global_variables_initializer())
np.array_equal(sess.run(conv), sess.run(conv))
```
`Out[2]: False`

**Comment by aselle** [comment]
@zheng-xq, Do you have any thoughts on this. I made a slightly more elaborate test case that computes relative error between the gpu and cpu version and two runs of the cpu version. @jnjaby's fear that gpu is not deterministic seems due to non-determinism in reduction (probably), because I only see 1e-5 error between two runs on the gpu. But I see a ~5770. relative  error between the gpu vs cpu. Could you please take a look:

```python
import tensorflow as tf
import numpy as np

np.random.seed(1234)
conv_ = np.random.randn(10, 7, 7, 56)

with tf.device("/cpu:0"):
    bottom = tf.constant(conv_, dtype=tf.float32)
    weight = tf.get_variable("weight_cpu", [9, 9, 1, 56], initializer=tf.random_normal_initializer(0, 0.001))
    bias = tf.get_variable("bias_cpu", initializer=np.zeros(1, dtype=np.float32))

    conv = tf.nn.conv2d_transpose(bottom, weight, [10, 19, 19, 1], [1, 3, 3, 1], padding='SAME')
    conv_cpu = tf.nn.bias_add(conv, bias)


with tf.device('/gpu:0'):
    bottom = tf.constant(conv_, dtype=tf.float32)
    weight = tf.get_variable("weight_gpu", [9, 9, 1, 56], initializer=tf.random_normal_initializer(0, 0.001))
    bias = tf.get_variable("bias_gpu", initializer=np.zeros(1, dtype=np.float32))

    conv = tf.nn.conv2d_transpose(bottom, weight, [10, 19, 19, 1], [1, 3, 3, 1], padding='SAME')
    conv_gpu = tf.nn.bias_add(conv, bias)


sess = tf.Session()
sess.run(tf.global_variables_initializer())
cpu_a=sess.run(conv_cpu)
gpu_a=sess.run(conv_gpu)
gpu_b=sess.run(conv_gpu)

def rel_error(a,ref):
  return np.max(np.abs((ref-a)/ref))
print ('relerror gpu_a vs cpu %f relerror gpu_b vs cpu 2 %f'%(rel_error(gpu_a, cpu_a), rel_error(gpu_b, cpu_a)))
print ('relerror gpu_a vs. gpu_b %f '%(rel_error(gpu_a, gpu_b)))
```

**Comment by jnjaby** [comment]
@aselle because the weight initialized by CPU and GPU are different. I try to initialize weights with another `tf.constant`. However, this time GPU is deterministic while CPU is not. That really confuses.
```python
import tensorflow as tf
import numpy as np

np.random.seed(1234)
conv_ = np.random.randn(10, 7, 7, 56)
weight_ = np.random.randn(9, 9, 1, 56)

with tf.device("/cpu:0"):
    bottom = tf.constant(conv_, dtype=tf.float32)
    weight = tf.constant(weight_, dtype=tf.float32, name="weight_cpu")
    bias = tf.get_variable("bias_cpu", initializer=np.zeros(1, dtype=np.float32))

    conv = tf.nn.conv2d_transpose(bottom, weight, [10, 19, 19, 1], [1, 3, 3, 1], padding='SAME')
    conv_cpu = tf.nn.bias_add(conv, bias)


with tf.device('/gpu:0'):
    bottom = tf.constant(conv_, dtype=tf.float32)
    weight = tf.constant(weight_, dtype=tf.float32, name="weight_gpu")
    bias = tf.get_variable("bias_gpu", initializer=np.zeros(1, dtype=np.float32))

    conv = tf.nn.conv2d_transpose(bottom, weight, [10, 19, 19, 1], [1, 3, 3, 1], padding='SAME')
    conv_gpu = tf.nn.bias_add(conv, bias)


sess = tf.Session()
sess.run(tf.global_variables_initializer())
cpu_a=sess.run(conv_cpu)
gpu_a=sess.run(conv_gpu)
gpu_b=sess.run(conv_gpu)

def rel_error(a,ref):
  return np.max(np.abs((ref-a)/ref))

print ('relerror gpu_a vs cpu %f \nrelerror gpu_b vs cpu  %f'%(rel_error(gpu_a, cpu_a), rel_error(gpu_b, cpu_a)))
print ('relerror gpu_a vs. gpu_b %f '%(rel_error(gpu_a, gpu_b)))

print (np.array_equal(sess.run(conv_cpu), sess.run(conv_cpu)))
print (np.array_equal(sess.run(conv_gpu), sess.run(conv_gpu)))
```

**Comment by zheng-xq** [comment]
I would agree with first making the initialization the same. One common way is to random initialize the weights in numpy, and use that in both paths.

**Comment by aselle** [comment]
This version shows no relative error between any of the implementations, so I am thinking there is no bug
```python
import tensorflow as tf
import numpy as np

np.random.seed(1234)
conv_ = np.random.randn(10, 7, 7, 56)

weight = np.random.uniform(-1.,1., (9,9,1,56)).astype(np.float32)
bias = np.random.uniform(-.1,.1, (1)).astype(np.float32)


with tf.device("/cpu:0"):
    bottom = tf.constant(conv_, dtype=tf.float32)
    #weight = tf.get_variable("weight_cpu", [9, 9, 1, 56], initializer=tf.random_normal_initializer(0, 0.001))
    #bias = tf.get_variable("bias_cpu", initializer=np.zeros(1, dtype=np.float32))

    conv = tf.nn.conv2d_transpose(bottom, weight, [10, 19, 19, 1], [1, 3, 3, 1], padding='SAME')
    conv_cpu = tf.nn.bias_add(conv, bias)


with tf.device('/gpu:0'):
    bottom = tf.constant(conv_, dtype=tf.float32)
    #weight = tf.get_variable("weight_gpu", [9, 9, 1, 56], initializer=tf.random_normal_initializer(0, 0.001))
    #bias = tf.get_variable("bias_gpu", initializer=np.zeros(1, dtype=np.float32))

    conv = tf.nn.conv2d_transpose(bottom, weight, [10, 19, 19, 1], [1, 3, 3, 1], padding='SAME')
    conv_gpu = tf.nn.bias_add(conv, bias)


sess = tf.Session()
sess.run(tf.global_variables_initializer())
cpu_a=sess.run(conv_cpu)
gpu_a=sess.run(conv_gpu)
gpu_b=sess.run(conv_gpu)
```

**Comment by jnjaby** [comment]
@aselle The last case is consistent. I make another test case output `True` and `False` respectively, showing different result between `tf.constant` and `tf.Variables`. Commonly, however, the weights should remain the same after initialized. So let's clarify my questions:
- Where is the non-determinism produced, Tensorflow or other libraries of GPU?
- What is the difference between `constant` Tensor and `variable` Tensor?

```python
import tensorflow as tf
import numpy as np

np.random.seed(1234)
conv_ = np.random.randn(10, 7, 7, 56)
weight = np.random.uniform(-1.,1., (9,9,1,56)).astype(np.float32)

with tf.device('/gpu:0'):
    bottom = tf.constant(conv_, dtype=tf.float32)
    biases = tf.get_variable("bias_gpu", initializer=np.zeros(1, dtype=np.float32))

    weight_con = tf.constant(weight, dtype=tf.float32)
    weight_var = tf.get_variable("weight_gpu", [9, 9, 1, 56], initializer=tf.random_normal_initializer(0, 0.001))

    conv_con = tf.nn.conv2d_transpose(bottom, weight_con, [10, 19, 19, 1],[1, 3, 3, 1],padding='SAME')
    conv_var = tf.nn.conv2d_transpose(bottom, weight_var, [10, 19, 19, 1],[1, 3, 3, 1],padding='SAME')

    conv_con = tf.nn.bias_add(conv_con, biases)
    conv_var = tf.nn.bias_add(conv_var, biases)

sess = tf.Session()
sess.run(tf.global_variables_initializer())

print (np.array_equal(sess.run(conv_con), sess.run(conv_con)))
print (np.array_equal(sess.run(conv_var), sess.run(conv_var)))
```

**Comment by ali01** [comment]
@zheng-xq, are you able to answer @jnjaby's questions above? Thanks!

**Comment by zheng-xq** [comment]
There are some known differences between CPU and GPU convolutions. Yang is working with Benoit to resolve them.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by ajbouh** [comment]
Any updates here? I keep having to switch back and forth between an old version of TensorFlow and the current one. Is addressing this a priority for the team?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by yzhwang** [comment]
Please refer to https://github.com/tensorflow/tensorflow/issues/14601#issuecomment-361734250 for updates. Once we are using the updated version of eigen, you can verify if the bug is fixed.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ajbouh** [comment]
I could be wrong, but I as of today, I don't think this issue has been addressed in the nightly wheel available over here: http://ci.tensorflow.org/view/tf-nightly/job/tf-nightly-mac/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3,label=mac-slave/lastSuccessfulBuild/artifact/pip_test/whl/tf_nightly-1.head-py3-none-any.whl

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yzhwang, @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yzhwang, @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Zealoe** [comment]
Sorry for bothering you, but are there any updates now? Thanks very much! @yzhwang  @benoitsteiner

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yzhwang, @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yzhwang, @benoitsteiner: It has been 62 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 77 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 92 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 107 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 122 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 137 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @robieta: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
Please check with the latest version of TensorFlow. Feel free to reopen if the issues still persists.
